cpp/path_module/algorithm/path.cpp

```diff
@@ -304,115 +346,90 @@ void Path::Expand(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result,
-      return;
-    }
-  }
-  try {
-    if (visited_nodes.at(node) <= hop_count) {
-      return;
+void Path::PathSubgraph::InsertNode(const mgp::Node &node, int64_t hop_count, LabelBools &label_bools) {
+  if (helper_.FilterNodes(hop_count == 0)) {
+    if (helper_.AreLabelsValid(label_bools)) {
+      to_be_returned_nodes_.AppendExtend(mgp::Value(node));
     }
-  } catch (const std::out_of_range &e) {
-    // it's okay, the node is not in visited nodes map
-    if (!is_start || config.At("minLevel").ValueInt() != 1) {
-      if ((labelFilterSets.end_list.empty() && labelFilterSets.termination_list.empty()) ||
-          IsLabelListed(node, labelFilterSets.end_list) || IsLabelListed(node, labelFilterSets.termination_list)) {
-        to_be_returned_nodes.AppendExtend(mgp::Value(node));
-      }
-    }
-  }
-  visited_nodes.insert({node, hop_count});
-  if (IsLabelListed(node, labelFilterSets.termination_list)) {
     return;
   }
-  for (const auto in_rel : node.InRelationships()) {
-    if (RelFilterAllows(config, in_rel.Type(), true)) {
-      VisitNode(in_rel.From(), visited_nodes, false, config, hop_count + 1, labelFilterSets, to_be_returned_nodes);
-    }
-  }
-  for (const auto out_rel : node.OutRelationships()) {
-    if (RelFilterAllows(config, out_rel.Type(), false)) {
-      VisitNode(out_rel.To(), visited_nodes, false, config, hop_count + 1, labelFilterSets, to_be_returned_nodes);
-    }
+
+  if (!visited_.contains(node.Id().AsInt())) {
+    to_be_returned_nodes_.AppendExtend(mgp::Value(node));
   }
 }
 
-void SetConfig(mgp::Map &config) {
-  auto value = config.At("maxLevel");
-  if (value.IsNull()) {
-    config.Insert("maxLevel", mgp::Value(int64_t(-1)));
-  }
-  value = config.At("relationshipFilter");
-  if (value.IsNull()) {
-    config.Insert("relationshipFilter", mgp::Value(mgp::List()));
-  }
-  value = config.At("labelFilter");
-  if (value.IsNull()) {
-    config.Insert("labelFilter", mgp::Value(mgp::List()));
-  }
-  value = config.At("filterStartNode");
-  if (value.IsNull()) {
-    config.Insert("filterStartNode", mgp::Value(false));
-  }
-  value = config.At("minLevel");
-  if (value.IsNull()) {
-    config.Insert("minLevel", mgp::Value(int64_t(0)));
+mgp::List Path::PathSubgraph::BFS() {
+  std::queue<Pair> queue;
+  std::unordered_set<int64_t> visited;
+
+  for (const auto &node : start_nodes_) {
+    queue.push({node, 0});
+    visited.insert(node.Id().AsInt());
   }
-  if (!(config.At("minLevel").IsInt() && config.At("maxLevel").IsInt() && config.At("relationshipFilter").IsList() &&
-        config.At("labelFilter").IsList() && config.At("filterStartNode").IsBool()) ||
-      (config.At("minLevel").ValueInt() != 0 && config.At("minLevel").ValueInt() != 1)) {
-    throw mgp::ValueException(
-        "The config parameter needs to be a map with keys and values in line with the documentation.");
+
+  while (!queue.empty()) {
+    auto pair = queue.front();
+    queue.pop();
+
```

Comment:
Copy paste previous comment:
This is an optimisation that can sometimes help a lot. If a node is unvisited you know that the current path you're on is the shortest path since you are running BFS so it's always the correct choice. If you don't mark it now you'll just add a lot of nodes in the queue that will be skipped over later because when you get to them the first path that expanded to them will already visit them and render them useless.

So thats why we want to add it before expanding to the node.